### PR TITLE
[client] Make copy-files work with stdin and stdout

### PR DESCRIPTION
**Pull request by MaximDanilov**
#578 Now working with stdin and stdout if template was setted from command line. Seems to be not a good solution, but if you have some thought how to optimaze, would be glad to hear it :-).

### CODE REVIEW

**Comment by MaximDanilov** [comment]
@Saviq Please review

**Comment by MaximDanilov** [comment]
@Saviq @townsend2010 Please review this request. I don't familiar with travis-ci, so I can't understand the error I received

**Comment by Saviq** [comment]
Hi @MaximDanilov sorry for the travis errors, we're having infrastructure issues. I'll trigger a try run that will at least give us one real result.

bors try

Overall, it's not great that we have to write a temporary file just to send it up to the instance. Just imagine if the file is several GBs. 
I'd much rather see this open an `scp` session to the remote and stream the contents from stdin. Think you could look into this? @townsend2010 would you have any pointers on how to approach this?

**Comment by MaximDanilov** [comment]
Hi @Saviq , I thought about that and it's better to stream into the remote file directly. I have already tried to not use temprorary file like using QTextStream directly in push_file method of scp client. And to make this solution work I need to this problems
1. What size of file would be on the remote instance? Because it doesn't work if I passed a zero or large size of file. It's seems that it is important to set correct size.

**Comment by Saviq** [comment]
>     1. What size of file would be on the remote instance? Because it doesn't work if I passed a zero or large size of file. It's seems that it is important to set correct size.

Oh... That's not great. Unless you can convince `scp` to accept a streamed file (I'd need to read into whether that's at all possible), maybe we could wrap `tee file` or `cat > file` inside the instance… Not as elegant, but if it's not possible to use `scp`… The temporary file isn't elegant either.

Let us know what you can find, thanks!

We're thinking of running a dedicated agent inside instances, so in the long run we could handle it this way.

**Comment by townsend2010** [comment]
Right, `scp` itself is not capable of reading from `stdin` and thus, the `libssh` api for scp can't either.  Given this constraint, we either accept the tmp file solution, come up with a whole new solution in general, or not support stdin/stdout type operations.

**Comment by townsend2010** [comment]
BTW, the sftp protocol can accept streaming from stdin, and there is an `sftp` api in `libssh`, so perhaps long term, switching to using the `sftp` api is the right thing to do.

**Comment by MaximDanilov** [comment]
I'll try to do smth with sftp api. Not guarentee that it would be fast, but I'll try. Thanks for the tips.

**Comment by MaximDanilov** [comment]
@townsend2010 @Saviq Hi. I made streaming file using sftp api. Please review it.  
And also wrote some code for push_file and pull_file methods. If you don't need it now, I can comment or remove it.

**Comment by townsend2010** [comment]
Hey @MaximDanilov,

Thanks for this!  I'll review soon:)

**Comment by Saviq** [comment]
Hey @MaximDanilov have you signed [the Ubuntu CLA](https://www.ubuntu.com/legal/contributors) yet? It's required so we can accept your contribution. Thanks!

**Comment by MaximDanilov** [comment]
Hi @Saviq , now should be ok with that :-)

**Comment by townsend2010** [comment]
@MaximDanilov,

Sorry for taking so long to get to reviewing this, but wanted to let you know that I haven't forgotten about you:)  I'll give you some feedback soon.

**Comment by Saviq** [comment]
Hey @MaximDanilov, CI rejected this because of low unit test coverage, would you be able to add tests for this?

**Comment by MaximDanilov** [comment]
Hello @Saviq , I'll try to do it on this weekends. Can you provide me some example? :-)

**Comment by ricab** [comment]
Hi @MaximDanilov, we have our tests [here](https://github.com/CanonicalLtd/multipass/tree/master/tests). They use googletest and googlemock. After building normally, you can run them by doing `make test`. The test log will then be in `<build_folder>/Testing/Temporary/LastTest.log`. You can also run the test binary directly: `<build_folder>/bin/multipass_tests` to follow execution live and in color ;) Thanks for contributing!

**Comment by ricab** [comment]
Actually I just noticed that, depending on much gtest likes your TERM variable, the color bit may require a `GTEST_COLOR=yes` env var (or the corresponding cmd line flag).

**Comment by townsend2010** [comment]
Hey @MaximDanilov,

So, you'll need to add tests in two places.

The first is in `test_cli_client.cpp` and will need to add tests for the handling of the `-`.  These tests are indicated by the `copy_files_cmd_*` tests.

The second set of tests will be the need to mock the sftp client handling you have added.  You can get ideas based on the scp mocking in `mock_scp.[h|cpp]` and `test_scp_client.cpp`.

I also noticed an issue in the `-` handling for `stdout`, but I will add that to the review.

**Comment by townsend2010** [comment]
Well, I can't seem to reproduce what I _thought_ I had seen before, so nevermind:)

**Comment by MaximDanilov** [comment]
Hello @townsend2010. Have written some tests by now, but sometimes don't understand behavior of libssh api. When I try to use sftp_new from tests, I get sigfault exception. Do not know what to do? Do you have any suggestions?

**Comment by townsend2010** [comment]
Hi @MaximDanilov,

Could you please show your code so I can see what you've done so far?

Thanks!

**Comment by MaximDanilov** [comment]
Okey. I'll commit tomorrow. Thanks for reply :-)

**Comment by MaximDanilov** [comment]
Hello @townsend2010, I have problems with throws_when_failed_to_init test. I do not understand, why i get sigfault on sftp_new.

**Comment by townsend2010** [comment]
Hi @MaximDanilov,

Thanks!  I'll have a look soon.

**Comment by townsend2010** [comment]
Hi @MaximDanilov,

Just an update...I've been looking at this and ran it through `gdb` and found that the `libssh` calls are not getting mocked like we expect.  This means the calls are going through the actual `libssh` calls and of course, things are not in the state that `libssh` expects and it crashes.

That said, I haven't figured out why they aren't getting mocked though, so I'm still digging at this.

File: src/ssh/sftp_client.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+void mp::SFTPClient::push_file(const std::string& source_path, const std::string& destination_path)
+{
+    auto full_destination_path = full_destination(destination_path, mp::utils::filename_for(source_path));
+    SFTPUPtr sftp{make_sftp_session(*ssh_session)};
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp push] init failed", sftp_init);
+
+    QFile source(QString::fromStdString(source_path));
+    const auto size{source.size()};
+    auto file_handle = sftp_open(sftp.get(), full_destination_path.c_str(), O_WRONLY|O_CREAT|O_TRUNC, file_mode);
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp push] failed", sftp_get_error);
```
[the remark this answers is not in the document]
Comment (MaximDanilov) [reply src/ssh/sftp_client.cpp]: +

File: src/ssh/sftp_client.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+void mp::SFTPClient::push_file(const std::string& source_path, const std::string& destination_path)
+{
+    auto full_destination_path = full_destination(destination_path, mp::utils::filename_for(source_path));
+    SFTPUPtr sftp{make_sftp_session(*ssh_session)};
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp push] init failed", sftp_init);
+
+    QFile source(QString::fromStdString(source_path));
+    const auto size{source.size()};
+    auto file_handle = sftp_open(sftp.get(), full_destination_path.c_str(), O_WRONLY|O_CREAT|O_TRUNC, file_mode);
```
[the remark this answers is not in the document]
Comment (MaximDanilov) [reply src/ssh/sftp_client.cpp]: +

File: src/ssh/sftp_client.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    std::array<char, 65536u> data;
+
+    auto full_destination_path = full_destination(destination_path, filename);
+    QFile destination(QString::fromStdString(full_destination_path));
+    if (!destination.open(QIODevice::WriteOnly))
+        throw std::runtime_error(
+            fmt::format("[sftp pull] error opening file for writing: {}", destination.errorString().toStdString()));
+
+    auto file_handle = sftp_open(sftp.get(), destination_path.c_str(), O_RDONLY, file_mode);
+    SSH::throw_on_error(sftp, *ssh_session, "[sftp pull] failed", sftp_get_error);
```
[the remark this answers is not in the document]
Comment (MaximDanilov) [reply src/ssh/sftp_client.cpp]: +

File: include/multipass/ssh/sftp_client.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2018 Canonical, Ltd.
```
[the remark this answers is not in the document]
Comment (MaximDanilov) [reply include/multipass/ssh/sftp_client.h]: +

File: src/ssh/sftp_client.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2018 Canonical, Ltd.
```
[the remark this answers is not in the document]
Comment (MaximDanilov) [reply src/ssh/sftp_client.cpp]: +

**Comment by MaximDanilov** [comment]
Hello @townsend2010, I've finally wrote some test. Please review this :-)

**Comment by townsend2010** [comment]
@MaximDanilov,

The `pull_throws_on_sftp_get_stat_failed` test is segfaulting for some reason in CI, but not locally.  We'll need to look into this.

**Comment by MaximDanilov** [comment]
@townsend2010, Hello :-) Do you find the problem why `pull_throws_on_sftp_get_stat_failed` sigfault on remote machine? Can I help you?

**Comment by townsend2010** [comment]
@MaximDanilov,

Sorry to be silent on this.  I looked into the segfault some, but I'm not able to reproduce on my machine.  This leads me to believe there is some race going on.  I haven't spent any more time trying to figure it out.

**Comment by Saviq** [comment]
I reproduced the crash in the snap build environment, [where travis runs it](https://github.com/CanonicalLtd/multipass/blob/66ba339306f1f67534d867bef4cc38b1ee4c41f0/.travis.yml#L57).

```
#0  0x0000000001591736 in multipass::SFTPClient::pull_file (this=0x7fffffffdfc0, source_path="foo", destination_path="bar") at /root/parts/multipass/src/src/ssh/sftp_client.cpp:124
        sftp = std::unique_ptr<sftp_session_struct> containing 0x232ad80
        file_attributes = 0x0
        size = 18371595
        filename = <error: Cannot access memory at address 0x19>
        total = 0
        data = {
          _M_elems = "\200\337\376\377\377\177\000\000\003\000\000\000\000\000\000\000foo", '\000' <repeats 45 times>, "\300\337\376\377\377\177\000\000\003\000\000\000\000\000\000\000foo", '\000' <repeats 29 times>, "\360\337\376\377\377
\177\000\000\003\000\000\000\000\000\000\000foo", '\000' <repeats 39717 times>...}
        full_destination_path = <error: Cannot access memory at address 0x45>
        destination = <incomplete type>
        raw_ptr = 0x0
        file_handle = std::unique_ptr<sftp_file_struct> containing 0x19
        r = 0
#1  0x00000000013c55b6 in SFTPClient_pull_throws_on_sftp_get_stat_failed_Test::TestBody (this=0x2345280) at /root/parts/multipass/src/tests/test_sftp_client.cpp:154
        gtest_caught_expected = false
        gtest_msg = {value = 0x1b348d9 ""}
        sftp = {ssh_session = std::unique_ptr<multipass::SSHSession> containing 0x23188d0}
        _150 = {_func = @0x2273f40,
          _oldFunc = {<std::_Maybe_unary_or_binary_function<sftp_session_struct*, ssh_session_struct*>> = {<std::unary_function<ssh_session_struct*, sftp_session_struct*>> = {<No data fields>}, <No data fields>}, <std::_Function_base> = {
static _M_max_size = 16, static _M_max_align = 8, _M_functor = {_M_unused = {_M_object = 0x7fffffffe070, _M_const_object = 0x7fffffffe070, _M_function_pointer = 0x7fffffffe070,
                  _M_member_pointer = (void (std::_Undefined_class::*)(std::_Undefined_class * const)) 0x7fffffffe070, this adjustment 20801277}, _M_pod_data = "p\340\377\377\377\177\000\000\375f=\001\000\000\000"},
              _M_manager = 0x13cba13 <std::_Function_base::_Base_manager<(anonymous namespace)::SFTPClient::SFTPClient()::<lambda(ssh_session)> >::_M_manager(std::_Any_data &, const std::_Any_data &, std::_Manager_operation)>},
            _M_invoker = 0x13cb9d5 <std::_Function_handler<sftp_session_struct*(ssh_session_struct*), (anonymous namespace)::SFTPClient::SFTPClient()::<lambda(ssh_session)> >::_M_invoke(const std::_Any_data &, <unknown type in /root/parts
/multipass/build/bin/multipass_tests, CU 0x933c5e, DIE 0x9839e6>)>}}
        _151 = {_func = @0x2273f80, _oldFunc = {<std::_Maybe_unary_or_binary_function<int, sftp_session_struct*>> = {<std::unary_function<sftp_session_struct*, int>> = {<No data fields>}, <No data fields>}, <std::_Function_base> = {
              static _M_max_size = 16, static _M_max_align = 8, _M_functor = {_M_unused = {_M_object = 0x7ffff68b1f6c <sftp_init>, _M_const_object = 0x7ffff68b1f6c <sftp_init>, _M_function_pointer = 0x7ffff68b1f6c <sftp_init>,
                  _M_member_pointer = (void (std::_Undefined_class::*)(std::_Undefined_class * const)) 0x7ffff68b1f6c <sftp_init>}, _M_pod_data = "l\037\213\366\377\177\000\000\000\000\000\000\000\000\000"},
              _M_manager = 0x11864d2 <std::_Function_base::_Base_manager<int (*)(sftp_session_struct*)>::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation)>},
            _M_invoker = 0x1186497 <std::_Function_handler<int (sftp_session_struct*), int (*)(sftp_session_struct*)>::_M_invoke(std::_Any_data const&, sftp_session_struct*&&)>}}
        _152 = {_func = @0x2273fe0,
          _oldFunc = {<std::_Maybe_unary_or_binary_function<sftp_attributes_struct*, sftp_session_struct*, char const*>> = {<std::binary_function<sftp_session_struct*, char const*, sftp_attributes_struct*>> = {<No data fields>}, <No data
fields>}, <std::_Function_base> = {static _M_max_size = 16, static _M_max_align = 8, _M_functor = {_M_unused = {_M_object = 0x7ffff68b82d6 <sftp_stat>, _M_const_object = 0x7ffff68b82d6 <sftp_stat>,
                  _M_function_pointer = 0x7ffff68b82d6 <sftp_stat>, _M_member_pointer = (void (std::_Undefined_class::*)(std::_Undefined_class * const)) 0x7ffff68b82d6 <sftp_stat>},
                _M_pod_data = "ւ\213\366\377\177\000\000\000\000\000\000\000\000\000"},
              _M_manager = 0x11869fd <std::_Function_base::_Base_manager<sftp_attributes_struct* (*)(sftp_session_struct*, char const*)>::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation)>},
            _M_invoker = 0x11869a8 <std::_Function_handler<sftp_attributes_struct* (sftp_session_struct*, char const*), sftp_attributes_struct* (*)(sftp_session_struct*, char const*)>::_M_invoke(std::_Any_data const&, sftp_session_struct*
&&, char const*&&)>}}
#2  0x00000000014dfd66 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (object=0x2345280, method=&virtual testing::Test::TestBody(), location=0x1b507b3 "the test body")
    at /root/parts/multipass/src/3rd-party/grpc/third_party/googletest/googletest/src/gtest.cc:2424
No locals.
```

The full traceback is [attached](https://github.com/CanonicalLtd/multipass/files/3176640/gdb.txt).

@MaximDanilov can you please rebase your work on master? We've renamed the command to `transfer` so there's a small conflict here.

**Comment by townsend2010** [comment]
Hi @MaximDanilov,

We've let this PR sit way too long.  Would it be ok with you if we take over this PR and fix it up?  Of course, you'll get the credit for the work you've already done:)

**Comment by MaximDanilov** [comment]
Hi @Saviq @townsend2010. Yes,  it would be ok if you'll take this PR and fix it. It's my first experience working open source and I'm glad that I helped you :-).If you need my help with another issues, I will be glad to help you.

**Comment by townsend2010** [comment]
Hi @MaximDanilov,

Great, I don't want to have to prolong your pain any more on this:)  We really appreciate your contribution and if you see any other bugs/enhancements you'd like to try, then please don't hesitate.

Best regards!

**Comment by townsend2010** [comment]
I'm closing this particular PR, but I've opened a new PR at #789.

**Pull request closed without merging**